commcare_connect/opportunity/visit_import.py

```diff
@@ -208,17 +208,48 @@ def get_status_by_visit_id(dataset) -> dict[int, VisitValidationStatus]:
     return status_by_visit_id, reason_by_visit_id
 
 
+def get_data_by_visit_id(dataset) -> dict[int, any]:
+    headers = [header.lower() for header in dataset.headers or []]
+    if not headers:
+        raise ImportException("The uploaded file did not contain any headers")
+
+    visit_col_index = _get_header_index(headers, VISIT_ID_COL)
+    status_col_index = _get_header_index(headers, STATUS_COL)
+    reason_col_index = _get_header_index(headers, REASON_COL)
+    justification_col_index = _get_header_index(headers, JUSTIFICATION_COL, required=False)
+    data_by_visit_id = defaultdict(dict)
+    invalid_rows = []
+    for row in dataset:
+        row = list(row)
+        visit_id = str(row[visit_col_index])
+        status_raw = row[status_col_index].lower().strip().replace(" ", "_")
+        try:
+            data_by_visit_id[visit_id]["status"] = VisitValidationStatus[status_raw]
+        except KeyError:
+            invalid_rows.append((row, f"status must be one of {VisitValidationStatus.values}"))
+        if status_raw == VisitValidationStatus.rejected.value:
+            data_by_visit_id[visit_id]["reason"] = str(row[reason_col_index])
+        if justification_col_index > 0:
+            data_by_visit_id[visit_id]["justification"] = str(row[justification_col_index])
+
+    if invalid_rows:
+        raise ImportException(f"{len(invalid_rows)} have errors", invalid_rows)
+    return data_by_visit_id
+
+
 def get_imported_dataset(file, file_format):
     if file_format == "csv":
         file = codecs.iterdecode(file, "utf-8")
     imported_data = Dataset().load(file, format=file_format)
     return imported_data
 
 
-def _get_header_index(headers: list[str], col_name: str) -> int:
+def _get_header_index(headers: list[str], col_name: str, required=True) -> int:
     try:
         return headers.index(col_name)
     except ValueError:
+        if not required:
+            return -1
```

Comment:
`-1` is a valid index, it might be better to return something like `None` that would raise an error if its improperly used. I see you check for this, so its not blocking, but might be safer in the future

Reply:
yeah, the type system was very unhappy with None so I had to resort to -1 and added a check for it.